H-976: Allow switching pages/entities from public to private

benwerner01 opened the pull request:
## 🌟 What is the purpose of this PR?

<!-- Explain, at a high level, what this does and why. -->
<!-- Use the 'What does this change?' section to list more specific implementation details. -->

This PR adds the ability to toggle a page or entity to "Public" and "Private".

## 🔗 Related links

<!-- Add links to any context it is worth capturing (e.g. Issues, Discussions, Discord) -->
<!-- Mark any links which are not publicly accessible as _(internal)_ -->
<!-- Don't rely on links to explain the PR, especially internal ones: use the sections above -->

- H-976

## 🔍 What does this change?

<!-- Use a bullet list to explain your changes in more detail, if it would be helpful. -->
<!-- If applicable, link to the specific commit.-->

See commits.

## Pre-Merge Checklist 🚀

### 🚢 Has this modified a publishable library?

<!-- Confirm you have taken the necessary action to record a changeset or publish a change, as appropriate -->
<!-- Tick AT LEAST ONE box and delete the rest. Do not delete this section! see libs/README.md for info on publishing -->

This PR:

- [x] does not modify any publishable blocks or libraries, or modifications do not need publishing

### 📜 Does this require a change to the docs?

<!-- If this adds a user facing feature or modifies how an existing feature is used, it likely needs a docs change. -->
<!-- Tick ONE box and delete the rest. Do not delete this section! -->

The changes in this PR:

- [x] are internal and do not require a docs change

### 🕸️ Does this require a change to the Turbo Graph?

<!-- If this adds or moves an existing package, modifies `scripts` in a `package.json`, it likely needs a turbo graph change. -->
<!-- Tick ONE box and delete the rest. Do not delete this section! -->

The changes in this PR:

- [x] do not affect the execution graph

## ⚠️ Known issues

<!-- Are there known issues / intentionally omitted functionality? Flag them here to save reviewers doing so -->

## 🐾 Next steps

<!-- Are there are planned/suggested follow ups which are related but won't be done in this PR? -->

## 🛡 What tests cover this?

<!-- What automated tests cover this? Existing ones? New ones? None? -->

None

## ❓ How to test this?

<!-- Tell reviewers how they can test the functionality -->

⚠️ This PR requires backend changes to be applied to main, so using the Vercel deployment risks breaking the production environment

1. Go to a page or an entity, and click "Share" in the top right
2. Toggle the entity to "Public", and refresh the page to ensure it has persisted correctly
3. Toggle back to "Private" and ensure it has persisted correctly

## 📹 Demo

<!-- Add a screenshot or video showcasing your work -->

https://github.com/hashintel/hash/assets/42802102/6fc44f49-919e-4b98-aeb0-32c9c657f95c

nonparibus commented:
@benwerner01 Can we show the icon (e.g. lock/globe/etc) next to the word 'Share' (left-side) so it's obvious without clicking in what the state of the page/entity is? I've updated the designs [in Figma](https://www.figma.com/file/gydVGka9FjNEg9E2STwhi2/HASH?type=design&node-id=1503%3A124874&mode=design&t=ZABnbcq628qWxpnk-1) (_internal link_) to reflect this.

benwerner01 commented:
> Can we show the icon (e.g. lock/globe/etc) next to the word 'Share' (left-side) so it's obvious without clicking in what the state of the page/entity is? I've updated the designs [in Figma](https://www.figma.com/file/gydVGka9FjNEg9E2STwhi2/HASH?type=design&node-id=1503%3A124874&mode=design&t=ZABnbcq628qWxpnk-1) (internal link) to reflect this.

Sure, addressed this in https://github.com/hashintel/hash/pull/3377/commits/1f5be19c634526d44cd2408bc9afce00b1840fc9

Now looks like:
<img width="444" alt="image" src="https://github.com/hashintel/hash/assets/42802102/cd1bd659-fca8-44db-97fa-d6a7a1a0da72">

<img width="433" alt="image" src="https://github.com/hashintel/hash/assets/42802102/502b4723-41d2-46c3-a49b-472ebaf15b2b">

CiaranMn commented on apps/hash-api/src/graphql/resolvers/knowledge/entity/entity.ts:
```diff
@@ -355,3 +372,168 @@ export const inferEntitiesResolver: ResolverFn<
   return status.contents[0];
 };
+
+const parseGqlOwnerInput = (ownerInput: OwnerInput) => {
```
I appreciate you aren't using this yet, but why go through the process of having the frontend need to know if it's an `accountId` or an `accountGroupId`, and then parse it back to a uuid here again, rather than just passing a `subjectId` or however-named uuid straight through?

CiaranMn replied:
One alternative is to make the input have some kind of type enum, which would have the advantage of avoiding the `public: true` input, which looks a bit weird when removing the relationship. Something like:
```graphql
enum SubjectType {
  Public
  Account
  AccountGroup
}

type ViewerInput {
  type: SubjectType!
  subjectId: OwnedById // or some scalar that combines AccountId and AccountGroupId
}

type EditorInput {
  type: SubjectType!
  subjectId: OwnedById!
}
```

benwerner01 replied:
I would also prefer using an enum, but to my understanding enums are not supported when defining GQL inputs (see [this explanation](https://github.com/graphql/graphql-wg/blob/main/rfcs/InputUnion.md)).

I can probably unify `accountId` and `accountGroupId` in a `subjectId`, but because unions cannot be expressed I would need to introduce some scalar that represents `AccountId | AccountGroupId` so that the argument is typed correctly. Would you still prefer this?

benwerner01 replied:
I stumbled upon `@oneof` whilst looking into this, and it seems to be supported by our GQL setup. This would mean I can convert the types to:

```GraphQL
input OwnerInput @oneOf {
    accountId: AccountId
    accountGroupId: AccountGroupId
  }

  input EditorInput @oneOf {
    accountId: AccountId
    accountGroupId: AccountGroupId
  }

  input ViewerInput @oneOf {
    accountId: AccountId
    accountGroupId: AccountGroupId
    public: Boolean
  }
```

And could get rid of the `parseGqlXInput` methods I made as the generated types will enforce at least one is provided:

<img width="808" alt="image" src="https://github.com/hashintel/hash/assets/42802102/b143a604-934a-4b7b-b9bb-26781f92d9bd">

CiaranMn replied:
> I would also prefer using an enum, but to my understanding enums are not supported when defining GQL inputs (see [this explanation](https://github.com/graphql/graphql-wg/blob/main/rfcs/InputUnion.md)).

You can use an enum as an input type – you cannot use unions of types.

I'm still not clear why the caller needs to know whether they're passing an `accountId` or an `accountGroupId` at all... nothing in the process cares which it is. By making it part of the argument we are forcing callers to know (or to find out) whether a subject is a user or an account.

I didn't know about `@oneof`, that's helpful!

benwerner01 replied:
Apologies I misunderstood what you suggested, I've unified the `OwnerInput` and `ViewerInput` into a single `AuthorizationSubjectId` for the inputs in https://github.com/hashintel/hash/pull/3377/commits/cf2a02b592fbc335d69d57145d08bef994bef67e. I still couldn't figure out a better way than to make this a scalar (because of the lack of union support), and also had to re-introduce the `parseGqlAuthorizationViewerInput` method to ensure that either a subject Id is provided when `kind !== "Public"`.

CiaranMn replied:
Looks good, thank you!
